README.md
# ObjectListView [![.NET](https://github.com/Furtif/ObjectListView/actions/workflows/dotnet.yml/badge.svg)](https://github.com/Furtif/ObjectListView/actions)
######
##### This is a fork of [Phillip Piper **_(in sourceforge)_**](https://sourceforge.net/p/objectlistview/code)
#### [Readme](http://objectlistview.sourceforge.net/cs/index.html)
#### [UpStream](https://github.com/Furtif/ObjectListView/tree/upstream)
#### [Discution](https://sourceforge.net/p/objectlistview/discussion)
### [Credits](https://sourceforge.net/u/grammarian)


CHANGELOG.txt
.. -*- coding: UTF-8 -*-

.. _changelog:

Change Log
==========

2008-09-04 23:12 (#250) - setup.py
  - v1.2


2008-09-04 22:30 (#249) - ObjectListView/ObjectListView.py
  - Correct an incomplete comment


2008-09-04 22:30 (#248) - Examples/SqlExample.py
  - Correctly locate primary key when there is a WHERE clause


2008-09-04 22:30 (#247) - Examples/BatchedUpdateExample.py
  - Nicely format file size column


2008-09-04 22:29 (#246) - docs/whatsnew.rst, docs/index.rst, docs/recipes.rst
  - Final changes for v1.2


2008-09-02 23:25 (#245) - docs/features.rst, docs/.static/icon.ico, docs/whatsnew.rst, docs/listCtrlPrinter.rst, docs/conf.py, docs/majorClasses.rst, docs/index.rst
  - Updated in preparation for v1.2 release


2008-09-02 23:22 (#244) - Examples/UsingVirtualListExample.py
  - Simplified initial insertions (removed executemany)


2008-09-02 23:21 (#243) - ObjectListView/__init__.py
  - Added BatchedUpdate


2008-09-02 23:20 (#242) - Examples/BatchedUpdateExample.py
  - First version


2008-09-02 23:20 (#241) - ObjectListView/ObjectListView.py
  - Added BatchedUpdate adaptor
  - Improved speed of selecting and refreshing by keeping a map of objects to indicies
  - Added GetIndexOf()
  - Removed flicker from FastObjectListView.AddObjects() and RefreshObjects()


2008-08-31 23:09 (#240) - Examples/BatchedUpdateExample.py
  - Scanning now works


2008-08-31 20:58 (#239) - Examples/BatchedUpdateExample.py
  - First hand done layout


2008-08-31 18:07 (#238) - Examples/BatchedUpdateExample.py
  - Initial checkin


2008-08-28 22:50 (#237) - CHANGELOG.txt, docs/changelog.rst
  - Rebuilt change log


2008-08-28 22:41 (#235) - test/test_ObjectListView.py
  - Added filtering tests


2008-08-28 22:41 (#234) - ObjectListView/ObjectListView.py
  - Added GetObjects() and GetFilteredObjects()
  - Added resortNow parameter to SetSortColumn()


2008-08-28 22:39 (#233) - ObjectListView/Filter.py
  - Added Filter.Chain
  - Added text constructor parameter to TextSearch


2008-08-28 01:22 (#232) - ObjectListView/ObjectListView.py
  - Correct AddObjects() when a filter is in effect
  - Made RebuildGroups() public


2008-08-28 01:21 (#231) - Examples/Demo.py
  - Implement search controls on several tabs


2008-08-28 01:20 (#230) - ObjectListView/Filter.py
  - Make text search handle non-report views better


2008-08-27 23:59 (#229) - ObjectListView/ObjectListView.py, ObjectListView/__init__.py, ObjectListView/Filter.py
  - Filters work


2008-08-25 10:51 (#225) - Examples/Demo.py
  - Use AddObjects() for "Add 1000" commands


2008-08-25 10:50 (#224) - ObjectListView/ObjectListView.py
  - Added AddObjects()/RemoveObjects() and friends
  - Removed duplicate code when building/refreshing/adding objects
  - One step closer to secondary sort column support


2008-08-22 19:38 (#220) - docs/listCtrlPrinter.rst
  - Added formatting picture


2008-08-20 22:21 (#219) - Examples/Demo.py
  - Changed to use new properties on ListViewPrinter


2008-08-20 22:20 (#218) - ObjectListView/ListCtrlPrinter.py
  - Consistently use properties on ListCtrlPrinter (ReportFormat, PageFooter, PageHeader, Watermark and PrintData are now all properties)
  - Removed ListCtrlPrinter.PageHeader(), ListCtrlPrinter.PageFooter(), ListCtrlPrinter.Watermark(), since they are now replaced with properties (and make more sense that way)


2008-08-20 00:28 (#217) - docs/.static/icon.ico, docs/images/listctrlprinter-example2.png, docs/images/listctrlprinter-structure.png, docs/listCtrlPrinter.rst, docs/.templates/layout.html, Examples/Demo.py
  - Added lots of documentation about ListCtrlPrinter


2008-08-20 00:27 (#216) - ObjectListView/ListCtrlPrinter.py
  - Moved AlwaysCenter and CanWrap to BlockFormat
  - Improved docs


2008-08-18 10:04 (#214) - THANKS.txt
  - Added Werner Bruhin to THANKS


2008-08-18 10:03 (#213) - ObjectListView/ObjectListView.py, ObjectListView/__init__.py, ObjectListView/OLVEvent.py
  - Handle model objects that cannot be hashed
  - Added editing started and finished events


2008-08-18 10:02 (#212) - Examples/SqlExample.py
  - Reorganized code slightly


2008-08-18 00:37 (#211) - Examples/SqlExample.py
  - Initial checkin


2008-08-17 21:47 (#210) - ObjectListView/WordWrapRenderer.py
  - Second attempt at avoid bug in wordwrap module


2008-08-16 23:31 (#209) - ObjectListView/WordWrapRenderer.py
  - Allow truncated text to be vertically aligned


2008-08-16 23:24 (#208) - ObjectListView/ListCtrlPrinter.py
  - Use RunningBlockPusher to simplify code
  - Allow truncated strings to be vertically aligned


2008-08-16 22:58 (#207) - ObjectListView/ListCtrlPrinter.py
  - Centralize cell width calculation (again)
  - Gracefully handle substitutions that fail


2008-08-16 22:55 (#206) - Examples/Demo.wxg, Examples/Demo.py
  - All control changes on ListCtrlPrinting now update the preview


2008-08-16 10:23 (#205) - ObjectListView/WordWrapRenderer.py
  - Avoid bug in wordwrap module
  - use DCClipper
  - Simplified some code


2008-08-16 09:47 (#204) - ObjectListView/ListCtrlPrinter.py
  - Column width is now calculated by the column headers only
  - Added ListCtrlPrinter.GetPrintData()
  - Make sure print data is destroyed after printing
  - Remove print statements


2008-08-16 09:38 (#203) - ObjectListView/ObjectListView.py
  - Added ensureVisible parameter to SelectObject()


2008-08-13 00:09 (#199) - ObjectListView/ObjectListView.py, ObjectListView/__init__.py, ObjectListView/OLVPrinter.py, ObjectListView/ListCtrlPrinter.py
  - Allow text to be vertically aligned in cells
  - Improved some docs
  - Renamed OLVPrinter to be ListCtrlPrinter


2008-08-13 00:07 (#198) - Examples/Demo.wxg, Examples/Demo.py
  - Changed OLVPrinter to be ListCtrlPrinter


2008-08-13 00:06 (#197) - docs/images/listctrlprinter-example1.png, docs/images/grouplist-example1.png, docs/listCtrlPrinter.rst, docs/.templates/layout.html, docs/faq.rst, docs/index.rst, docs/.static/listCtrlPrinter-icon.png, docs/.static/majorClasses-icon.png

Examples/OwnerDrawnEditor.py
# -*- coding: ISO-8859-1 -*-
#----------------------------------------------------------------------------
# Name:         OwnerDrawnEditor.py
# Author:       Phillip Piper
# Created:      9 April 2008
# Copyright:    (c) 2008 by Phillip Piper
# RCS-ID:       $Id: timectrl.py 32483 2005-02-28 18:45:13Z RD $
# License:      wxWindows license
#----------------------------------------------------------------------------
# Change log:
# 2008/04/09  JPP   Initial version complete
#----------------------------------------------------------------------------
# To do:

"""
This module holds two owner drawn combo boxes.

FontFaceComboBox allows the user to choose a font face. It displays each font face by
drawing a string in that font.

ColourComboBox allows the user to choose a colour, It displays each colour by drawing
the colour's name on a background of that colour.
"""

import wx
import wx.combo
import wx.lib.colourdb

class Bucket:
    """
    General purpose data container
    """
    def __init__(self, **kwargs):
        self.__dict__.update(kwargs)

#======================================================================

class FontFaceComboBox(wx.combo.OwnerDrawnComboBox):
    """
    FontFaceComboBox allows the user to choose a font face.

    Interface methods:

        GetValue()
            Returns a string with the name of the font face chosen by the user.

        SetValue(fontOrString)
            Set the font face that is chosen.

    Public attributes:

        fontHeight
            When string is drawn in the popup window, what should be the pixel size of the font?

        maximumPopupWidth
            When the popup window is showen, what is the maximum width that should be allowed?
            The actual width is taken from the width of the widest rendered string.

        evenRowBackground
            In the popup window, what will be the background colour of even rows?

        oddRowBackground
            In the popup window, what will be the background colour of odd rows?
    """

    def __init__(self, *args, **kwargs):
        """
        Initialize a font face combobox.

        In addition to the public attributes describes as public attributes,
        the constructor understands the following parameters:

        extendedText
            This string will be added to the face name to give an indication
            of how the font appears. Typically a panagram.
        """
        self.fontHeight = kwargs.pop("fontHeight", 18)
        self.maximumPopupWidth = kwargs.pop("maximumPopupWidth", 400)
        self.evenRowBackground = kwargs.pop("evenRowBackground", wx.WHITE)
        self.oddRowBackground = kwargs.pop("oddRowBackground", wx.Colour(191, 239, 255))
        extendedText = kwargs.pop("extendedText", " - Big fjords vex quick waltz nymph")

        # Build the data behind the control
        fe = wx.FontEnumerator()
        fe.EnumerateFacenames()
        self.fontInfo = [ Bucket(name=x, display=x+extendedText) for x in sorted(fe.GetFacenames(), key=unicode.lower) ]

        kwargs['style'] = kwargs.get("style", 0) | wx.CB_READONLY
        kwargs["choices"] = [x.name for x in self.fontInfo]
        wx.combo.OwnerDrawnComboBox.__init__(self, *args, **kwargs)

        # Fill in some other information that is better to precalculate.
        # Measuring text has to be done after the control is initialized
        measuringDC = wx.ClientDC(self)
        for x in self.fontInfo:
            x.font = wx.FFont(self.fontHeight, wx.FONTFAMILY_DEFAULT, face=x.name)
            x.extent = measuringDC.GetFullTextExtent(x.display, font=x.font)

    def SetValue(self, value):
        """
        Set the selected value to be the given value, which can be a string or a font
        """
        if isinstance(value, wx.Font):
            value = value.GetFaceName()

        wx.combo.OwnerDrawnComboBox.SetValue(self, value or "")

    def OnDrawItem(self, dc, rect, item, flags):
        if item == wx.NOT_FOUND:
            return

        bucket = self.fontInfo[item]
        if (flags & wx.combo.ODCB_PAINTING_CONTROL) == wx.combo.ODCB_PAINTING_CONTROL:
            fontSize = min(self.fontHeight, max(10, self.GetSize().GetHeight() - 9))
            dc.SetFont(wx.FFont(fontSize, wx.FONTFAMILY_DEFAULT, face=bucket.name))
            display = bucket.name
        else:
            dc.SetFont(bucket.font)
            display = bucket.display
        dc.DrawText(display, rect.x + 3, rect.y + ((rect.height - dc.GetCharHeight()) / 2))

    def OnDrawBackground(self, dc, rect, item, flags):
        # If the item is selected, or we are painting the combo control itself, then use
        # the default rendering.
        if flags & (wx.combo.ODCB_PAINTING_CONTROL | wx.combo.ODCB_PAINTING_SELECTED):
            wx.combo.OwnerDrawnComboBox.OnDrawBackground(self, dc, rect, item, flags)
            return

        # Otherwise, draw every other background with different colour.
        if item & 1:
            backColour = self.oddRowBackground
        else:
            backColour = self.evenRowBackground
        dc.SetBrush(wx.Brush(backColour))
        dc.SetPen(wx.Pen(backColour))
        dc.DrawRectangleRect(rect)

    def OnMeasureItem(self, item):
        (width, height, descent, externalLeading) = self.fontInfo[item].extent
        return min(self.fontHeight * 2, height + descent)

    def OnMeasureItemWidth(self, item):
        (width, height, descent, externalLeading) = self.fontInfo[item].extent
        return min(self.maximumPopupWidth, width)


#======================================================================

class ColourComboBox(wx.combo.OwnerDrawnComboBox):
    """
    ColourComboBox allows the user to choose a colour.

    Interface methods:

        GetValue()
            Returns a wx.Colour chosen by the user.

        SetValue(colour)
            Set the colour that is chosen.

    Public attributes:

        popupRowHeight
            When the popup window is showen, what is the height of each row?
    """

    def __init__(self, *args, **kwargs):
        self.popupRowHeight = kwargs.pop("popupRowHeight", 24)

        # Collect the colours with different values
        self.colourList = []
        lastColour = None
        for x in wx.lib.colourdb.getColourInfoList():
            colour = x[1:]
            if colour != lastColour:
                self.colourList.append(Bucket(name=x[0], colour=wx.Colour(*colour), intensity=sum(colour)))
                lastColour = colour

        kwargs['style'] = kwargs.get('style', 0) | wx.CB_READONLY
        kwargs['choices'] = [x.name for x in self.colourList]
        wx.combo.OwnerDrawnComboBox.__init__(self, *args, **kwargs)

    GetStringValue = wx.combo.OwnerDrawnComboBox.GetValue

    def GetValue(self):
        """
        Return a wx.Colour that is selected
        """
        strValue = self.GetStringValue()
        for x in self.colourList:
            if x.name == strValue:
                return x.colour
        return None

    def SetValue(self, value):
        """
        Set the selected value to be the given value, which can be a string or a colour
        """
        if isinstance(value, wx.Colour):
            for x in self.colourList:
                if x.colour == value:
                    value = x.name
                    break

        wx.combo.OwnerDrawnComboBox.SetValue(self, value or "")

    def OnDrawItem(self, dc, rect, item, flags):
        if item == wx.NOT_FOUND:
            return

        bucket = self.colourList[item]

        if flags & wx.combo.ODCB_PAINTING_SELECTED:
            dc.DrawRectangleRect(rect)
        else:
            if bucket.intensity > (255+255+255)*2/3:
                dc.SetTextForeground(wx.BLACK)
            else:
                dc.SetTextForeground(wx.WHITE)

        dc.DrawText(bucket.name, rect.x + 3, rect.y + ((rect.height - dc.GetCharHeight()) / 2))

    def OnDrawBackground(self, dc, rect, item, flags):
        # If the item is selected, then use the default rendering.
        # Otherwise, draw each row with its own colour as the background
        if flags & wx.combo.ODCB_PAINTING_SELECTED:
            wx.combo.OwnerDrawnComboBox.OnDrawBackground(self, dc, rect, item, flags)
        else:
            backColour = self.colourList[item].colour
            dc.SetBrush(wx.Brush(backColour))
            dc.SetPen(wx.Pen(backColour))
            dc.DrawRectangleRect(rect)

    def OnMeasureItem(self, item):
        return self.popupRowHeight

#======================================================================
# Testing only

if __name__ == '__main__':

    class MyFrame(wx.Frame):
        def __init__(self, *args, **kwds):
            kwds["style"] = wx.DEFAULT_FRAME_STYLE
            wx.Frame.__init__(self, *args, **kwds)

            self.panel = wx.Panel(self, -1)
            self.cbFont = FontFaceComboBox(self.panel, -1)
            self.cbFont2 = FontFaceComboBox(self.panel, -1, fontHeight=12, extendedText="Junk MTV quiz graced by fox whelps",
                                        maximumPopupWidth=250, evenRowBackground=wx.Colour(191, 229, 245), oddRowBackground=wx.Colour(255, 248, 220))
            self.cbColour = ColourComboBox(self.panel, -1)
            self.cbColour2 = ColourComboBox(self.panel, -1, popupRowHeight=14)

            sizer_2 = wx.BoxSizer(wx.VERTICAL)
            sizer_2.Add(self.cbFont, 1, wx.ALL, 4)
            sizer_2.Add(self.cbFont2, 1, wx.ALL, 4)
            sizer_2.Add(self.cbColour, 1, wx.ALL, 4)
            sizer_2.Add(self.cbColour2, 1, wx.ALL, 4)
            self.panel.SetSizer(sizer_2)

            sizer_1 = wx.BoxSizer(wx.VERTICAL)
            sizer_1.Add(self.panel, 1, wx.ALL, 4)
            self.SetSizer(sizer_1)
            self.Layout()

            self.cbFont.Bind(wx.EVT_COMBOBOX, self.HandleFontComboBox)
            self.cbFont2.Bind(wx.EVT_COMBOBOX, self.HandleFontComboBox2)
            self.cbColour.Bind(wx.EVT_COMBOBOX, self.HandleColourComboBox)
            self.cbColour2.Bind(wx.EVT_COMBOBOX, self.HandleColourComboBox2)

            wx.CallAfter(self.run)

        def HandleFontComboBox(self, evt):
            print self.cbFont.GetValue()

        def HandleFontComboBox2(self, evt):
            print self.cbFont2.GetValue()

        def HandleColourComboBox(self, evt):
            print self.cbColour.GetValue()

        def HandleColourComboBox2(self, evt):
            print self.cbColour2.GetValue()

        def run(self):
            self.cbColour.SetValue("CORNSILK")

    app = wx.PySimpleApp(0)
    wx.InitAllImageHandlers()
    frame_1 = MyFrame(None, -1, "")
    app.SetTopWindow(frame_1)
    frame_1.Show()
    app.MainLoop()
